### DISCUSSION TITLE
蓝牙连接问题

### DISCUSSION THREAD

**Discussion by a276994712** [wh201906/SerialTest#46] [Q&A]
我选了一个低功耗的蓝牙模块，该蓝牙模块是一个透传从机，但是实在不知道用你的软件如何连上我的模块收发数据。现状是我用卖家自带的手机测试app能够正常连接并收发数据，用serialtest不知道如何连接。我把蓝牙测试app附件源码传上来。方便指导下我们怎么用你的app连接我的模块，我的模块是PW02，测试软件主要连接PW02,PW01两种型号的蓝牙。附件是我测试软件截图。附件pwiot2.zip 是app 该测试软件源码。帮忙指导下如何操作，感激不尽，也可邮箱联系276994712@qq.com
[pwiot2.zip](https://github.com/wh201906/SerialTest/files/13470644/pwiot2.zip)

![截图2](https://github.com/wh201906/SerialTest/assets/21340270/9d96cc13-4a19-4e9c-8988-8fa67e1f99b7)
![截图1](https://github.com/wh201906/SerialTest/assets/21340270/2325671e-a533-4de4-aa86-9d93b9cee9f1)

**Comment by wh201906** [comment]
从代码上来看，PW01的接收/发送服务UUID都是0000fff0-0000-1000-8000-00805f9b34fb，接收/发送特征值UUID都是0000fff6-0000-1000-8000-00805f9b34fb。
PW02的的接收/发送服务UUID都是0783b03e-8535-b5a0-7140-a304d2495cb7，接收特征值UUID是0783b03e-8535-b5a0-7140-a304d2495cb8，发送特征值UUID是0783b03e-8535-b5a0-7140-a304d2495cba。
在已知某个设备的这四个UUID的情况下直接填到SerialTest里面然后点打开就行了

**Reply by a276994712** [reply]
是选择BLE中心设备（Central），这个里面去连接，填入你说的这4个特征值吗，感谢回答，我一会试试

**Reply by wh201906** [reply]
> 是选择BLE中心设备（Central），这个里面去连接

是的

> 填入你说的这4个特征值吗

每个设备对应的4个UUID分别是发送服务UUID，发送特征值UUID，接收服务UUID，接收特征值UUID。你需要根据上面的内容把对应的UUID填进去。

**Reply by a276994712** [reply]
![截图4](https://github.com/wh201906/SerialTest/assets/21340270/bd583e83-1612-4ab7-87e3-496d59de6a9a)
对应都填了还是无法建立连接
![截图3](https://github.com/wh201906/SerialTest/assets/21340270/2de8b657-4483-426a-86d2-6e77a2856abb)

**Reply by wh201906** [reply]
第一项的MAC地址填了吗？

**Reply by a276994712** [reply]
自动填的，填过了，不知道这个是否有日志呢，我看代码也是按你说的建立连接的但是实在搞不懂，问题出在哪

**Reply by wh201906** [reply]
PW01能连上吗？

**Reply by a276994712** [reply]
我只买了PW02,我手里的测试app，能连PW02的

**Reply by wh201906** [reply]
你可以把手机接电脑上，用Logcat看一下日志

**Reply by wh201906** [reply]
你还可以试试点最右边“打开”之前先点一下左边的“断开”

**Reply by a276994712** [reply]
能简单说下怎么连接电脑抓日志么，android 的开发不是特别熟悉，或则我网上搜了研究下。刚点了左边的断开，在点右边的打开，还是不行

**Reply by wh201906** [reply]
你需要在网上下载一个adb程序，然后在命令提示符或者Powershell里面运行`adb logcat`相关指令。过程不算复杂但在这里讲不太清楚，你最好还是自己在网上搜一下

**Reply by a276994712** [reply]
正在下载了，有结果我马上给你反馈

**Reply by wh201906** [reply]
辛苦了

**Reply by a276994712** [reply]
[logcat_test.txt](https://github.com/wh201906/SerialTest/files/13476016/logcat_test.txt)
[logcat_test1.txt](https://github.com/wh201906/SerialTest/files/13476018/logcat_test1.txt)
附件收集了两次我连接的日志，有空就看把。辛苦你了

**Reply by a276994712** [reply]
对比我发的日志，你这个app和我测试的app，有什么思路么，大佬，难道连接上有些区别吗？

**Reply by wh201906** [reply]
没发现很明显的错误，我准备自己买个PW02来试试

**Reply by a276994712** [reply]
我这有个开发版，我可以给你邮寄过来

**Reply by a276994712** [reply]
加我微信吧，17748760679，我把开发板给你邮寄过来，或则我给你买个

**Reply by wh201906** [reply]
没事，应该是我这边代码有问题，我就自费买了

**Reply by a276994712** [reply]
没关系，主要我最近对你这个有需求，也算是帮我自己了。方便加个微信吗

**Reply by a276994712** [reply]
【淘宝】https://m.tb.cn/h.5PQWMdX?tk=J0FWWVKng8L CZ0001 「PW02 蓝牙模块串口透传 超小封装 冰点低功耗BLE5.2工业级迷你ble」     我这个链接买的，价格到不贵，主要导师催得紧，我可以顺丰给你邮过来。

**Reply by wh201906** [reply]
从广东那边运过来应该挺快的

**Reply by a276994712** [reply]
好的，有进展麻烦第一时间通知，感激不尽！

**Reply by wh201906** [reply]
找到原因了。PW02的写特征值只有WriteNoResponse属性，但我当时做BLE功能的时候参考的蓝牙模块写特征值有Write属性，因此我的所有写入操作都是针对于含有Write属性的特征值来进行的。
要想完整支持这两种情况可能要大改一下代码，不过微调一下还是可以支持PW02。你需要在哪个平台上使用呢？如果要得急的话我可以微调完然后编译好发给你

**Reply by a276994712** [reply]
我主要目前在android平台上，测试使用，目前就是采集了数据想通过蓝牙看下实时绘图，然后继续调试我的电路板，主要我设计的时候，没有再预留其他串口了，直连蓝牙模块了。如果您方便的话，帮忙弄一版我自己测试下，感激不尽。

**Reply by wh201906** [reply]
试试这个吧，我测下来能用
[android-build-debug.zip](https://github.com/wh201906/SerialTest/files/13501841/android-build-debug.zip)
压缩包解压之后是APK文件，里面用的是调试签名，所以有些手机可能没法直接点开安装，这种情况你把手机连上电脑，打开USB调试，然后运行`adb install <APK路径>`就可以了。并且由于是调试签名，和我在F-Droid上以及GitHub Releases里面的apk文件签名不符，因此你需要先卸载已有的SerialTest再安装。

**Reply by a276994712** [reply]
好的，我今天试下。

**Reply by wh201906** [reply]
> 主要我设计的时候，没有再预留其他串口了，直连蓝牙模块了。

如果只接收数据的话可以从USB转串口模块上面把Rx和GND焊到板子上的Tx和GND上，也就是蓝牙模块和USB转串口模块的Rx连在一起，共用板子上的一个Tx。大部分的芯片IO驱动能力还是能带动两个模块的。

**Reply by a276994712** [reply]
现在可以连接了，感谢
